### PR TITLE
Add ParametricAttention.v2

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This layer is an extension of the existing `ParametricAttention` layer, adding support for transformations (such as a non-linear layer) of the key representation. This brings the model closer to the paper that suggested it (Yang et al, 2016) and gave slightly better results in experiments.

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: thinc/tests/layers/test_layers_api.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 from thinc.api import Dropout, Model, NumpyOps, registry, with_padded
 from thinc.backends import NumpyOps
 from thinc.compat import has_torch
+from thinc.layers.relu import Relu
```
Comment (svlandeg) [inline thinc/tests/layers/test_layers_api.py:11]: Why is this import needed?

File: thinc/tests/layers/test_layers_api.py [hunk 2]
```diff
@@ -129,6 +130,8 @@ def assert_data_match(Y, out_data):
     ("MultiSoftmax.v1", {"nOs": (1, 3)}, array2d, array2d),
     # ("CauchySimilarity.v1", {}, (array2d, array2d), array1d),
     ("ParametricAttention.v1", {}, ragged, ragged),
+    ("ParametricAttention.v2", {}, ragged, ragged),
+    ("ParametricAttention.v2", {"key_transform": {"@layers": "Gelu.v1"}}, ragged, ragged),
```
Comment (svlandeg) [inline thinc/tests/layers/test_layers_api.py:134]: We should also have a test to check that the `key_transform` layer is actually inserted in the model, as opposed to when it's `None`.

File: thinc/layers/parametricattention_v2.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        params={"Q": None},
+        dims={"nO": nO},
+        layers=layers,
+        refs=refs,
+    )
+
+
+def forward(model: Model[InT, OutT], Xr: InT, is_train: bool) -> Tuple[OutT, Callable]:
+    Q = model.get_param("Q")
+    key_transform = model.maybe_get_ref("key_transform")
```
Comment (svlandeg) [inline thinc/layers/parametricattention_v2.py:41]: With respect to the unit tests. I'm thinking that if we would somehow misspell this line, something like `model.maybe_get_ref("key_transformmm")` then our unit tests wouldn't catch this, nor would users - everything would still run as is as `key_transform` is allowed to be `None`. So let's think of ways we can test this more robustly.

File: thinc/layers/parametricattention_v2.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        dX += dX2
+        return Ragged(dX, dYr.lengths)
+
+    return Ragged(output, Xr.lengths), backprop
+
+
+def init(
+    model: Model[InT, OutT], X: Optional[InT] = None, Y: Optional[OutT] = None
+) -> None:
+    key_transform = model.maybe_get_ref("key_transform")
```
Comment (svlandeg) [inline thinc/layers/parametricattention_v2.py:61]: The other thing we should probably do, is to store this `"key_transform"` string in a variable at the top of this file.

**Review by svlandeg — COMMENTED** [review]
Looks good! I only had a few concerns about robustness & tests.

File: thinc/layers/parametricattention_v2.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    nO: Optional[int] = None
+) -> Model[InT, OutT]:
+    if key_transform is None:
+        layers = []
+        refs = {}
+    else:
+        layers = [key_transform]
+        refs = {"key_transform": cast(Optional[Model], key_transform)}
+
+    layers = [key_transform] if key_transform is not None else []
```
Comment (danieldk) [inline thinc/layers/parametricattention_v2.py:25]: I was thinking, all this custom handling could be avoided if we had an identity layer. If `key_transform` is `None`, we could just use the identity layer. Then we don't need any conditionals after that (also simplifies the testing al lot).

Comment (svlandeg) [reply thinc/layers/parametricattention_v2.py:25]: Good point! Isn't the `noop` basically an identity layer?

Comment (danieldk) [reply thinc/layers/parametricattention_v2.py:25]: It is, thanks :), I somehow looked over it because the standard terminology is identity.

Comment (danieldk) [reply thinc/layers/parametricattention_v2.py:25]: I'll update the PR after my current training run is done.

File: thinc/tests/layers/test_layers_api.py (hunk 1 above)
Comment (danieldk) [reply thinc/tests/layers/test_layers_api.py]: Removed.

File: thinc/tests/layers/test_layers_api.py (hunk 2 above)
Comment (danieldk) [reply thinc/tests/layers/test_layers_api.py:133]: Added a test that checks if the transform is added to the model.

File: thinc/layers/parametricattention_v2.py (hunk 3 above)
Comment (danieldk) [reply thinc/layers/parametricattention_v2.py]: This shouldn't occur anymore. Since the `noop` layer is now used, all the `maybe_get_ref`s are replaced by `get_ref`s.

File: thinc/layers/parametricattention_v2.py (hunk 4 above)
Comment (danieldk) [reply thinc/layers/parametricattention_v2.py]: Done.

File: website/docs/api-layers.md [hunk 6]
```diff
@@ -686,6 +686,44 @@ attention mechanism.
...
+<inline-list>
+
+- **Input:** <ndarray>Ragged</ndarray>
+- **Output:** <ndarray>Ragged</ndarray>
+- **Parameters:** <ndarray shape="nO,">Q</ndarray>
+
+</inline-list>
+
+A layer that uses the parametric attention scheme described by
+[Yang et al. (2016)](https://www.cs.cmu.edu/~./hovy/papers/16HLT-hierarchical-attention-networks.pdf).
```
Comment (adrianeboyd) [inline website/docs/api-layers.md:700]:
```suggestion
[Yang et al. (2016)](https://aclanthology.org/N16-1174/).
```

**Review by svlandeg — APPROVED** [review]
Definitely much cleaner implementation with the `noop` layer. Looks good to merge!
